src/runtime/virtcontainers/sandbox.go

```diff
@@ -825,6 +834,103 @@ func (s *Sandbox) coldOrHotPlugVFIO(sandboxConfig *SandboxConfig) (bool, error)
 	return coldPlugVFIO, nil
 }
 
+func (s *Sandbox) prepareInitdataMount(config *HypervisorConfig) error {
+	if len(config.Initdata) == 0 {
+		s.Logger().WithField("initdata", "No initdata provided. Skip prepare initdata device").Info("No initdata provided. Skip prepare loop device")
+		return nil
+	}
+
+	s.Logger().Info("Start to prepare initdata")
+	initdataWorkdir := filepath.Join(string(filepath.Separator), "/run/initdata", s.id)
+	initdataImagePath := filepath.Join(string(filepath.Separator), initdataWorkdir, "data.img")
+
+	err := os.MkdirAll(initdataWorkdir, 0755)
+	if err != nil {
+		s.Logger().WithField("initdata", "create squashfs image path").WithError(err)
+		return err
+	}
+
+	CreateFilesystem := func(d *disk.Disk, spec disk.FilesystemSpec) (filesystem.FileSystem, error) {
+		// find out where the partition starts and ends, or if it is the entire disk
+		var (
+			size, start int64
+		)
+
+		switch {
+		case spec.Partition == 0:
+			size = d.Size
+			start = 0
+		case d.Table == nil:
+			return nil, fmt.Errorf("cannot create filesystem on a partition without a partition table")
+		default:
+			partitions := d.Table.GetPartitions()
+			// API indexes from 1, but slice from 0
+			part := spec.Partition - 1
+			if spec.Partition > len(partitions) {
+				return nil, fmt.Errorf("cannot create filesystem on partition %d greater than maximum partition %d", spec.Partition, len(partitions))
+			}
+			size = partitions[part].GetSize()
+			start = partitions[part].GetStart()
+		}
+
+		switch spec.FSType {
+		case filesystem.TypeSquashfs:
+			return squashfs.Create(d.Backend, size, start, d.LogicalBlocksize)
+		default:
+			return nil, errors.New("unknown filesystem type requested")
+		}
+	}
+
+	// Create squashfs image file
+	var diskSize int64 = 10 * 1024 * 1024 // 10 MB
+
+	// TODO: Allow bigger initdata size
+	// Here we use a block size of 1MB, because there is a bug in the underlying
+	// library.
+	// This limits the size of initdata that can be processed normally to 1MB.
+	// See https://github.com/diskfs/go-diskfs/issues/282
```

Comment:
Interesting. We have limitations with the initdata size across the whole stack. Let's be sure to document this and try to improve in future iterations.

Reply:
Hopefully we can get https://github.com/diskfs/go-diskfs/pull/287 merged so we do not have such limit

Reply:
Merged and updated in this PR